File: 2024/day-04/day4.py
def process_input():
  with open("./day-04/input.txt", "r") as f:
    matrix = f.read().splitlines()
    rows = len(matrix)
    cols = len(matrix[0])
    return matrix, rows, cols

def check_part1(x, y, letters, index, dx, dy):
  if index == len(letters):
    return 1
  
  nx, ny = x + dx, y + dy
  if 0 <= ny < rows and 0 <= nx < cols:
    if matrix[ny][nx] == letters[index]:
      return check_part1(nx, ny, letters, index + 1, dx, dy)
  return 0

def solve_part1():
  directions = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, -1), (1, -1), (-1, 1)]
  res = 0

  for y in range(rows):
    for x in range(cols):
      if matrix[y][x] == 'X':
        for dx, dy in directions:
          res += check_part1(x, y, ['X', 'M', 'A', 'S'], 1, dx, dy)
  return res

def solve_part2():
  res = 0
  for r in range(rows):
    for c in range(cols):
      if matrix[r][c] == 'A':
        diagonal1 = False
        diagonal2 = False
        if 0 <= r + 1 < rows and 0 <= r - 1 < rows and 0 <= c + 1 < cols and 0 <= c - 1 < cols:
          diagonal1 = (matrix[r + 1][c + 1] == "M" and matrix[r - 1][c - 1] == "S") or \
            (matrix[r + 1][c + 1] == "S" and matrix[r - 1][c - 1] == "M")
          diagonal2 = (matrix[r + 1][c - 1] == "M" and matrix[r - 1][c + 1] == "S") or \
            (matrix[r + 1][c - 1] == "S" and matrix[r - 1][c + 1] == "M")
        if diagonal1 and diagonal2:
          res += 1
  return res

if __name__ == "__main__":
  matrix, rows, cols = process_input()
  print(solve_part1())
  print(solve_part2())

File: 2024/day-04/day4.go
package main

import (
  "bufio"
  "os"
  "fmt"
  "strings"
)

type Direction struct {
  dx int
  dy int
}

func process_input() [][]rune {
  f, err := os.Open("./day-04/input.txt")
  if err != nil {
    fmt.Println("Error opening file:", err)
    os.Exit(1)
  }
  defer f.Close()

  var matrix [][]rune
  scanner := bufio.NewScanner(f)
  for scanner.Scan() {
    line := strings.TrimSpace(scanner.Text())
    row := []rune(line)
    matrix = append(matrix, row) 
    // Can be slightly optimized by preallocating memory for matrix, requiring 
    // a prior search of cost O(m) for an m*n matrix to determine no. of rows.
  }
  if err := scanner.Err(); err != nil {
    fmt.Println("Error reading file:", err)
    os.Exit(1)
  }
  return matrix
}

func solvePart1(matrix [][]rune) int {
  var noOfRows int = len(matrix)
  var noOfCols int = len(matrix[0])

  var directions []Direction = []Direction{
    {1, 0},
    {-1, 0},
    {0, 1},
    {0, -1},
    {1, 1},
    {-1, -1},
    {1, -1},
    {-1, 1},
  }

  var result int = 0
  for y := 0; y < noOfRows; y++ {
    for x := 0; x < noOfCols; x++ {
      if matrix[y][x] != 'X' {
        continue
      }

      for _, direction := range directions {
        result += checkPart1(matrix, x, y, direction, 1)
      }
    }
  }
  return result
}

func checkPart1(matrix [][]rune, x, y int, direction Direction, index int) int {
  var noOfRows int = len(matrix)
  var noOfCols int = len(matrix[0])
  var letters []rune = []rune{'X', 'M', 'A', 'S'}

  if index == len(letters) {
    return 1
  }

  nx, ny := x + direction.dx, y + direction.dy

  if isWithinLimit(nx, ny, noOfRows, noOfCols) && matrix[ny][nx] == letters[index] {
    return checkPart1(matrix, nx, ny, direction, index + 1)
  }
  return 0
}

func solvePart2(matrix [][]rune) int {
  var noOfRows int = len(matrix)
  var noOfCols int = len(matrix[0])

  var result int = 0
  for y := 0; y < noOfRows; y++ {
    for x := 0; x < noOfCols; x++ {
      if matrix[y][x] != 'A' {
        continue
      }

      diagonal1 := isWithinLimit(y + 1, x + 1, noOfRows, noOfCols) &&
                   isWithinLimit(y - 1, x - 1, noOfRows, noOfCols) &&
                   ((matrix[y + 1][x + 1] == 'M' && matrix[y - 1][x - 1] == 'S') ||
                    (matrix[y + 1][x + 1] == 'S' && matrix[y - 1][x - 1] == 'M'))
      diagonal2 := isWithinLimit(y + 1, x - 1, noOfRows, noOfCols) &&
                   isWithinLimit(y - 1, x + 1, noOfRows, noOfCols) &&
                   ((matrix[y + 1][x - 1] == 'M' && matrix[y - 1][x + 1] == 'S') ||
                    (matrix[y + 1][x - 1] == 'S' && matrix[y - 1][x + 1] == 'M'))

      if diagonal1 && diagonal2 {
        result++
      }
    }
  }
  return result
}

func isWithinLimit(x, y, noOfRows, noOfCols int) bool {
  return x >= 0 && x < noOfCols && y >= 0 && y < noOfRows
}

func main() {
  var matrix [][]rune = process_input()
  fmt.Println("Part 1 Solution: ", solvePart1(matrix))
  fmt.Println("Part 2 Solution: ", solvePart2(matrix))
}

